src/shaka_scheme/system/base/Vector.cpp

```diff
@@ -0,0 +1,87 @@
+#include "shaka_scheme/system/base/Vector.hpp"
+
+#include "shaka_scheme/system/exceptions/IndexOutOfBoundsException.hpp"
+
+// I included this so I can create unspecified values
+#include "shaka_scheme/system/core/types.hpp"
+#include "shaka_scheme/system/base/Data.hpp"
+
+#include <cctype>
+#include <cstring>
+#include <utility>
+
+namespace shaka {
+
+Vector::Vector(std::size_t k) :
+    arr(new shaka::NodePtr[k]),
+    size(k) {
+  // Initialize all of the NodePtr to unspecified value
+  for (std::size_t i = 0; i < k; ++i) {
+    this->arr[i] = core::create_unspecified_node();
+  }
+}
+
+Vector::Vector(std::size_t k, NodePtr fill) :
+    arr(new NodePtr[k]),
+    size(k) {
+  // Initialize all of the nodes to a copy of the NodePtr value.
+  for (std::size_t i = 0; i < k; ++i) {
+    this->arr[i] = create_node(*fill);
+  }
+}
+
+Vector::Vector(std::initializer_list<NodePtr> l) :
+    arr(new NodePtr[l.size()]),
+    size(l.size()) {
+  // For each NodePtr in the initializer list,
+  // copy the value into the vector
+  std::size_t i = 0;
+  for (const auto& it : l) {
+    this->arr[i] = create_node(*it);
+    ++i;
+  }
+}
+Vector::Vector(const Vector& other) :
+    arr(new NodePtr[other.length()]),
+    size(other.length()) {
+  // Copy each of the elements from the other vector into this one.
+  for (std::size_t i = 0; i < size; ++i) {
+    this->arr[i] = create_node(*other.arr[i]);
+  }
+}
+
+Vector::Vector(Vector&& other) :
+    arr(std::move(other.arr)),

```

Comment:
The alternative approach is to say `arr(nullptr), size(0) { swap(*this, other); } ` However, your approach works too, although I do not believe the calls to `std::move` on `other.arr` and `other.size` are necessary, since arr is just a raw pointer, and size is just an unsigned number. `std::move` just casts these things to R values, but I am not aware that move semantics really apply to raw pointers, or numeric primitive types in general that is.